File: src-tauri/src/officellm_bridge.js
// OfficeLLM QuickJS Bridge
// Bridges officellm.open()/doc.call()/doc.save()/doc.close() to workspace.officellm()
// See ~/.officellm/skills/resources/QUICKJS_API_GUIDE.md for the full spec.

var officellm = (function() {
  function camelToKebab(s) {
    return s.replace(/[A-Z]/g, function(c) { return "-" + c.toLowerCase(); });
  }

  function camelToSnake(s) {
    return s.replace(/[A-Z]/g, function(c) { return "_" + c.toLowerCase(); });
  }

  // Convert camelCase JS params to kebab-case CLI params.
  // boolean true  -> empty string (bare flag, e.g. --dry-run)
  // boolean false -> omitted entirely
  // null/undefined -> omitted
  // everything else -> String(value)
  function convertParams(params) {
    var out = {};
    for (var key in params) {
      if (!params.hasOwnProperty(key)) continue;
      var val = params[key];
      var k = camelToKebab(key);
      if (typeof val === "boolean") {
        if (val) out[k] = "";
      } else if (val != null) {
        out[k] = String(val);
      }
    }
    return out;
  }

  // Call workspace.officellm() and parse the JSON result.
  function invoke(cmd, params) {
    var raw = workspace.officellm(cmd, params || {});
    try { return JSON.parse(raw); } catch (e) { return raw; }
  }

  function createSession() {
    return {
      call: function(command, params) {
        return invoke(command, convertParams(params || {}));
      },

      execute: function(ops, options) {
        var batch = { version: "1.0", ops: ops };
        if (options) {
          for (var k in options) {
            if (options.hasOwnProperty(k)) {
              batch[camelToSnake(k)] = options[k];
            }
          }
        }
        return invoke("execute", { "instructions-json": JSON.stringify(batch) });
      },

      save: function(path) {
        return invoke("save", path ? { path: path } : {});
      },

      close: function() {
        return invoke("close", {});
      }
    };
  }

  return {
    open: function(path) {
      invoke("open", { path: path });
      return createSession();
    },
    create: function(params) {
      invoke("create", convertParams(params || {}));
      return createSession();
    },
    call: function(command, params) {
      return invoke(command, convertParams(params || {}));
    }
  };
})();

File: src-tauri/src/officellm_bridge.lua
-- OfficeLLM Lua Bridge
-- Bridges officellm.open()/doc.call()/doc.save()/doc.close() to workspace.officellm()

local function camel_to_kebab(s)
    return s:gsub("%u", function(c) return "-" .. c:lower() end)
end

local function camel_to_snake(s)
    return s:gsub("%u", function(c) return "_" .. c:lower() end)
end

-- Convert camelCase Lua params to kebab-case CLI params.
-- boolean true  -> empty string (bare flag, e.g. --dry-run)
-- boolean false -> omitted entirely
-- nil           -> omitted
-- everything else -> tostring(value)
local function convert_params(params)
    local out = {}
    for key, val in pairs(params) do
        local k = camel_to_kebab(key)
        if type(val) == "boolean" then
            if val then out[k] = "" end
        elseif val ~= nil then
            out[k] = tostring(val)
        end
    end
    return out
end

-- Call workspace.officellm() and parse the JSON result.
local function invoke(cmd, params)
    local raw = workspace.officellm(cmd, params or {})
    local ok, result = pcall(json.decode, raw)
    if ok then return result else return raw end
end

local function create_session()
    return {
        call = function(command, params)
            return invoke(command, convert_params(params or {}))
        end,

        execute = function(ops, options)
            local batch = { version = "1.0", ops = ops }
            if options then
                for k, v in pairs(options) do
                    batch[camel_to_snake(k)] = v
                end
            end
            return invoke("execute", { ["instructions-json"] = json.encode(batch) })
        end,

        save = function(path)
            if path then
                return invoke("save", { path = path })
            else
                return invoke("save", {})
            end
        end,

        close = function()
            return invoke("close", {})
        end,
    }
end

officellm = {
    open = function(path)
        invoke("open", { path = path })
        return create_session()
    end,

    create = function(params)
        invoke("create", convert_params(params or {}))
        return create_session()
    end,

    call = function(command, params)
        return invoke(command, convert_params(params or {}))
    end,
}
